src/base/memory_monitor_sr.hpp

```diff
@@ -0,0 +1,266 @@
+  class MMM_heap_module : public MMM_module
+  {
+    public:
+      heap_stats (const char *name) : template_stats (name) {}
+
+      /* function */
+      virtual int aggregate_stats (MEMMON_MODULE_INFO *info);
+
+      ~heap_stats()
+      {
+	if (m_module_name)
+	  {
+	    delete[] m_module_name;
+	  }
+      }
+  };
+
+  class MMM_aggregater
+  {
+    public:
+      MMM_aggregater (memory_monitoring_manager *mmm)
+      {
+	this->m_memmon_mgr = mmm;
+      }
+
+      /*  backlink of memory_monitoring_manager class */
+      memory_monitoring_manager *m_memmon_mgr;
+
+      /* function */
+      int get_server_info (MEMMON_SERVER_INFO *info);
+      int get_module_info (MEMMON_MODULE_INFO *info);
+      int get_transaction_info (MEMMON_TRAN_INFO *info);
+
+      ~MMM_printer() {}
+  };
+
+  /* you have to add your modules in Memory Monitoring Manager(MMM) class */
+  class memory_monitoring_manager
+  {
+    public:
+      memory_monitoring_manager (const char *name)
+      {
+	this->m_server_name = new char[strlen (name) + 1];
+	strcpy (this->m_server_name, name);
+	aggregater = new MMM_aggregater (this);
+      }
+      /* variable */
+      char *server_name;
+      std::atomic<uint64_t> m_total_mem_usage;
+      MMM_aggregater *m_aggregater;
+
+      /* Your module class should add in this array with print function */
+      template_stats *m_module[MMM_MODULE_END] =
+      {
+	new template_stats(),				  /* dummy */
+	new heap_stats ("heap", heap_comp_info)
+      };
+
+      /* function */
+      int add_stat (THREAD_ENTRY *thread_p, MMM_STATS stat_name, uint64_t size);
+      int sub_stat (THREAD_ENTRY *thread_p, MMM_STATS stat_name, uint64_t size);
+      int resize_stat (THREAD_ENTRY *thread_p, MMM_STATS stat_name, uint64_t before_size, uint64_t after_size);
+      int move_stat (THREAD_ENTRY *thread_p, MMM_STATS before_stat_name, MMM_STATS after_stat_name, uint64_t size);
+      int aggregate_module_info (MEMMON_MODULE_INFO *info, int module_index);
+      int aggregate_tran_info (MEMMON_TRAN_INFO *info, int tran_count);
+      int get_module_index (char *name)
+      {
+	for (int i = 1; i <= MMM_MODULE_END; i++)
+	  {
+	    if (!strcmp (modules[i]->module_name, name))
+	      {
+		return i;
+	      }
+	  }
+	return 0;		// error case
+      }
+
+      static inline int MMM_COMP_IDX_FROM_COMP_IDX_MAP (int comp_idx_map_val)
+      {
+	return comp_info_map_val >> 16;
+      }
+
+      static inline int MMM_SUBCOMP_IDX_FROM_COMP_IDX_MAP (int comp_idx_map_val)
+      {
+	return comp_info_map_val & MMM_PARSE_MASK;
+      }
+
+      ~MMM()
+      {
+	delete[] m_server_name;
+	delete m_aggregater;
+	for (int i = 0; i < MMM_MODULE_END; i++)
+	  {
+	    delete m_module[i];
+	  }
+      }
+  };
+  extern MMM *MMM_global;
+#if defined(NDEBUG)
+  extern std::atomic<uint64_t> test_meminfo;
+#endif
+} // namespace cubperf
+
+//#endif // __cplusplus
+#endif
+#endif // _MEMORY_MONITOR_HPP_
```

Comment:
```suggestion
#endif // _MEMORY_MONITOR_SR_HPP_
```

Reply:
It has already been modified as hornetmj mentioned.